File: 13-roman-to-integer/13-roman-to-integer.py
class Solution(object):
    def romanToInt(self, s):
        """
        :type s: str
        :rtype: int
        """
        map = {"I": 1, "V":5, "X":10, "L":50, "C":100, "D":500, "M":1000, "Z":0}
        i, sum = 0, 0
        while i < len(s):
            if i+1 < len(s) and map[s[i]] < map[s[i+1]]:
                sum += map[s[i+1]] - map[s[i]]
                i += 2
            else:
                sum += map[s[i]]
                i += 1
        return sum
        

File: 13-roman-to-integer/13-roman-to-integer.java
class Solution {
    static Map<Character, Integer> map = new HashMap<>();
    static {
        map.put('M', 1000);
        map.put('D', 500);
        map.put('C', 100);
        map.put('L', 50);
        map.put('X', 10);
        map.put('V', 5);
        map.put('I', 1);  
    }
    
    public int romanToInt(String s) {
        int sum = 0;
        int i = 0;
        while (i < s.length()) {
            int curr = map.get(s.charAt(i));
            int next = (i+1 < s.length()) ? map.get(s.charAt(i+1)) : 0;
            if (curr < next) {
                sum += next - curr;
                i += 2;
            }
            else {
                sum += curr;
                i++;
            }
        }
        return sum;
    }
}